### PR TITLE
[knx] Add discovery service

### PR DESCRIPTION
**Pull request by holgerfriedrich**
This PR adds automatic discovery of KNX IP interfaces and routers.
It will create a thing in the inbox (for routers it might create two entries, if both routing and tunneling is announced).
Discovery uses multicast traffic to IP 224.0.23.12, port 3671.

### CODE REVIEW

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        if (scanFuture != null) {
+            scanFuture.cancel(false);
+            scanFuture = null;
+        }
+    }
+
+    @Override
+    protected void startBackgroundDiscovery() {
+        // only start once at startup
+        startScan();
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java:91]: You already start it in `activate()`.
The "background discovery" is meant to do a continuous scan in the background, so that the Inbox is filled even if the user does not trigger a scan. So I would rather expect a scheduled job here.

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+@Component(service = DiscoveryService.class, configurationPid = "discovery.knx")
+@NonNullByDefault
+public class KNXnetDiscoveryService extends AbstractDiscoveryService {
+    private final Logger logger = LoggerFactory.getLogger(KNXnetDiscoveryService.class);
+
+    private @Nullable Future<?> scanFuture = null;
+
+    public void activate() {
+        super.activate(null);
+        startScan();
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java:57]: You should actually only trigger a scan here, if background discovery is enabled.

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+    @Override
+    protected void stopBackgroundDiscovery() {
+        stopScan();
+    }
+
+    private synchronized void startDiscovery() {
+        try {
+            logger.debug("Starting KNXnet/IP discovery scan");
+            Thread.sleep(1000);
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java:102]: Why is this needed?

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                                .withProperty("port", "" + response.getControlEndpoint().getPort())
+                                .withRepresentationProperty("serialNumber").build());
+                    }
+                } else {
+                    logger.trace("Ignoring device {}", response);
+                }
+            }
+            logger.debug("Completed KNXnet/IP discovery scan");
+        } catch (Exception ex) {
+            logger.info("An error occurred during KNXnet/IP discovery {}", ex.getMessage(), ex);
```
Comment (kaikreuzer) [inline bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java:146]: I guess this could rather be a warning as it is an unexpected situation, isn't it?

**Review by kaikreuzer — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java (hunk 1 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java:91]: removed

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java (hunk 3 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java:102]: removed. It was not necessarily needed...

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java (hunk 4 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java:146]: Maybe. I consider this feature as less important for the normal operation of the binding, thus info could still be fine.
If you like, I can change it to a warning.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java:146]: The question is probably, what kind of exceptions could turn up here. Is it possible to catch more specific ones?
If the exception is something that should never happen, it should be logged as warning. If it is something that is expected to happen in certain setups (network constallations), then it can be info or even debug.

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java (hunk 2 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java]: this way?
```
    @Activate
    public void activate() {
        super.activate(null);
        if (isBackgroundDiscoveryEnabled()) {
            startScan();
        }
    }
```

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java (hunk 4 above)
Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java]: It is mainly about degrading gracefully, there were some unexpected exceptions coming from Calimero in the past. I changed it to warning, let's see what we observe.

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java (hunk 2 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java]: Yes.

File: bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java (hunk 1 above)
Comment (kaikreuzer) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java]: Removing it means you are now not reacting to a configuration change of "background discovery enabled" anymore - i.e. you are not supporting any background discovery.
You have two options now:
1. Either call `super()` in the constructor with `false` instead of `true` now or
2. schedule a recurring job here (and unschedule it again in `stopBackgroundDiscovery()`).

Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java]: What I want to archive is basically:
- at startup scan once (unless disabled)
- maybe start scanning if user switches from disabled to enabled
- at best have it triggerable from the OH console: ```discovery start knx```

I think a found the way using the methods from ```AbstractDiscoveryService```. Seems to work as intended.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java]: This means you do not want to support background discovery, i.e. you choose option (1) from above.
As a result, you should call `super` with `false` [here](https://github.com/openhab/openhab-addons/pull/16033/files#diff-68ecbf945b85d60351bc4142da436b4dc4d3a980e045763ac756cbcaeeba7c00R55) and not implement anything in the start/stop background discovery methods.

Comment (holgerfriedrich) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java]: Hm, I think the current implementation is a bit of both: I have done it as background discovery, but the scan is only running once.
The pro of this approach is:
- it can be completely disabled using ```discovery.knx:background=false```
- it can be triggered by ```discovery start knx```
- and it is running only once at startup

The alternative as you outlined above, use ```super(..., false)``` and start the scan from the ctor. Then we users cannot disable it.

Comment (kaikreuzer) [reply bundles/org.openhab.binding.knx/src/main/java/org/openhab/binding/knx/internal/discovery/KNXnetDiscoveryService.java]: Hm, ok, I see your point. An IP interface is unlikely to appear suddenly, so having a single scan at startup is usually enough and it is convenient for the user to find it in the inbox without having to manually trigger a scan for KNX things. So alright, let's go for this.

**Review by kaikreuzer — APPROVED** [review]
lgtm, thanks!
